### PR TITLE
V2 golangci lint

### PR DESCRIPTION
**Pull request by vincentserpoul**

### CODE REVIEW

File: scanner.go [hunk 1]
```diff
@@ -111,6 +111,7 @@ func scanWhitespace(b []byte) ([]byte, []byte) {
 	return b, b[len(b):]
 }
 
+//nolint:unparam
```
Comment (vincentserpoul) [inline scanner.go:117]: scanComment - result 0 ([]byte) is never used (unparam)
Not sure if it's temporar or permanent, so I left it here. It basically says the first result is always ignored

File: unmarshaler.go [hunk 2]
```diff
@@ -90,49 +94,63 @@ func (d *decoder) arrayIndex(append bool, v reflect.Value) int {
...
-	}
 	if err == nil {
-		err = d.strict.Error(p.data)
+		return d.strict.Error(p.data)
+	}
+
+	var e *decodeError
+	if errors.As(err, &e) {
+		return wrapDecodeError(p.data, e)
 	}
```
Comment (vincentserpoul) [inline unmarshaler.go:110]: probably the highlight of this linting session: casting errors should be done this way!

File: unmarshaler_test.go [hunk 3]
```diff
@@ -1171,11 +1238,15 @@ key3 = "value3"
...
 	fmt.Println(err.Error())
-	// Output: strict mode: fields in the document are missing in the target struct
 
-	details := err.(*toml.StrictMissingError)
+	var details *toml.StrictMissingError
+	if !errors.As(err, &details) {
+		panic(fmt.Sprintf("err should have been a *toml.StrictMissingError, but got %s (%T)", err, err))
+	}
+
 	fmt.Println(details.String())
```
Comment (vincentserpoul) [inline unmarshaler_test.go:1247]: Spelling mistake was preventing the complete test

File: unmarshaler_test.go [hunk 4]
```diff
@@ -1171,11 +1238,15 @@ key3 = "value3"
...
-	// Ouput:
+	// Output:
```
Comment (vincentserpoul) [inline unmarshaler_test.go:1248]: Spelling mistake was preventing the complete test

File: scanner.go (hunk 1 above)
Comment (pelletier) [reply scanner.go:117]: Not using it now, but we will when implementing the Document interface, which will allow to manipulate comments.

File: scanner.go [hunk 5]
```diff
@@ -100,27 +110,31 @@ func scanWhitespace(b []byte) ([]byte, []byte) {
...
 		}
 	}
-	return b, nil, nil
+
+	return b, nil
 }
 
+var errBasicLineNotTerminatedByQuote = errors.New(`basic string not terminated by "`)
+
+//nolint:godox
```
Comment (pelletier) [inline scanner.go:137]: The TODO can actually be removed

File: scanner.go [hunk 6]
```diff
@@ -142,22 +156,22 @@ func scanBasicString(b []byte) ([]byte, []byte, error) {
 		}
 	}
 
-	return nil, nil, fmt.Errorf(`basic string not terminated by "`)
+	return nil, nil, errBasicLineNotTerminatedByQuote
 }
 
+//nolint:godox
```
Comment (pelletier) [inline scanner.go:162]: Same here. Those TODOs were here when those functions were `parse*` functions in the initial v2 code, but now there are `scan*` and `decode*` functions, with the latter handling validation.

File: unmarshaler.go (hunk 2 above)
Comment (pelletier) [reply unmarshaler.go:110]: neat!

File: unmarshaler.go [hunk 7]
```diff
@@ -252,6 +285,7 @@ func (d *decoder) scopeWithArrayTable(x target, key ast.Iterator) (target, bool,
 		x, err = scopeSlice(true, x)
 	case reflect.Array:
 		x, err = d.scopeArray(true, x)
+	default:
```
Comment (pelletier) [inline unmarshaler.go:288]: That doesn't seem useful 😄

File: unmarshaler_test.go (hunk 4 above)
Comment (pelletier) [reply unmarshaler_test.go:1248]: 🤦🏻‍♂️ good catch

**Comment by pelletier** [comment]
Merged it. I'll delete the TODOs when I do a full pass on them.

File: unmarshaler.go (hunk 7 above)
Comment (vincentserpoul) [reply unmarshaler.go:288]: indeed... just to satisfy the linter. I hesitated between nolint and this, but i guess it's less bad this way :)

**Pull request merged**